### ISSUE TITLE
shell_pipe (`|`) chomps final newline

### ISSUE THREAD

**Issue by adiabatic** [helix-editor/helix#11570]
### Summary

When I select a line or more by pressing `x` one or more times and pipe the text through https://github.com/adiabatic/quote-educator, the output text has one fewer newline in it, seemingly only in Helix.

### Reproduction Steps

To demonstrate, here's a sequence of screenshots.

First, I have a text file with a line of text in it:

![0  start](https://github.com/user-attachments/assets/62a6000a-0bb6-4497-9666-5acc37ce6a9e)

Then, I press `Esc` and then `x` to select the entire line:

![1  x](https://github.com/user-attachments/assets/5b1d85bc-0651-4dd1-b807-e78292a09cac)

Then I press `|` and pipe the whole thing through `quote-educator`:

![2  pipe through quote educator](https://github.com/user-attachments/assets/a71c5a84-7202-4d89-8a0a-c1ee3a3705cd)

…and then the quote is curled, but the file has one less line in it:

![3  one line removed](https://github.com/user-attachments/assets/9e58c8ec-5365-4bb9-97c4-0f95d31af754)

I'm pretty sure `quote-educator` isn't chomping the final newline. Here I'm demonstrating how it preserves newlines when I'm running it on input fed into it from [fish](https://fishshell.com)'s `printf`:

![Screenshot 2024-08-25 at 2 06 03 AM](https://github.com/user-attachments/assets/17d47c4e-c05c-42e9-b163-e4927a6de682)

As you can see, it's not turning the second invocation into the same output as the first one (with its "`⏎`" indicator that no newline was output, but `fish` is adding one anyway).

### Helix log

<details><summary>~/.cache/helix/helix.log (actually not)</summary>

(There are no log entries from when I was doing the above demonstration, so I'm not preemptively providing logs. I'll still provide logs on request.)

</details>

### Platform

macOS

### Terminal Emulator

iTerm2

### Installation Method

brew

### Helix Version

helix 24.7 (079f5442)

**Comment by pascalkuthe** [comment]
The behavior is intentional. Most shell commands always append a trailing newline to their output so we strip the final newline of pipe output. There is no way to perfectly since different commands behave inconsistenly in this regard.
Our behavior matches from kakoune.

**Comment by adiabatic** [comment]
> The behavior is intentional. Most shell commands always append a trailing newline to their output so we strip the final newline of pipe output. There is no way to perfectly since different commands behave inconsistenly in this regard. Our behavior matches from kakoune.

Fair enough. But:

I [added an option to add an extra newline to the output in `quote-educator`](https://github.com/adiabatic/quote-educator/commit/03a211980b15717e59ad7e1db3123c27032d24be), then re-built the binary and copied it to `~/bin`.

It seems to do what it ought to in `fish`:

![0  works in fish](https://github.com/user-attachments/assets/6dcbb9b0-62a4-4bba-842c-4539c66ecace)

But when I set things up in Helix…

![1  write the line](https://github.com/user-attachments/assets/3f960553-cd33-4d5e-8631-55e7fcab6c1c)

…and then press `x` to select the whole line…

![2  press x](https://github.com/user-attachments/assets/6c74e669-32b7-4614-97fd-7eeb577869fd)

…and then run `quote-educator -n`…

![3  qen](https://github.com/user-attachments/assets/f3eee5f8-1119-411e-a749-a663d03efd2d)

…I'm down one newline anyway.

![4  snip](https://github.com/user-attachments/assets/59ea1702-d9e9-4e3f-927c-820bfa88335c)

I shouldn't need to append _two_ newlines to my output to get what I want, right?

----

What's more, it appears that final spaces are chomped, too. I tried just selecting `I'd ` by pressing `W`:

![5  W](https://github.com/user-attachments/assets/0d51ed38-4b50-4770-8e61-b24157f3c3ab)

But after running `quote-educator -n` on it, the space after `I’d` is gone:

![6  spacenapped](https://github.com/user-attachments/assets/3134cbd8-249e-4fcd-b36f-b26db737920b)

----

However, if there is no final space, there's no chomping. If I select `I'd` by typing `E`…

![7  E](https://github.com/user-attachments/assets/ddfd05ce-247f-495a-97eb-10b071ac47ee)

…then the space doesn't get eaten when the quote gets curled by running `quote-educator -n` on it:

![8  curl](https://github.com/user-attachments/assets/1ae0f58d-9e8b-4352-87f8-5436ccbbc110)

----

I [just verified that `quote-educator` doesn't swallow trailing spaces, either](https://github.com/adiabatic/quote-educator/commit/6cd01b261e4b93b8ec926ff084875dcf69d8aa1e).

Am I wrong, or is Helix doing something other than chomping a trailing newline coming back in from input that's come back in through a pipe? It _seems_ like it's doing the equivalent of pressing `d` if the final character is either `\n` or ` `.

Thanks for having a second look.

**Comment by krish-r** [comment]
I think this was already fixed in #11183 (immediately after 24.07 release), so can you try the latest master?

Related issue - #11177.

**Comment by adiabatic** [comment]
I did a `git clone` and then a `cargo build`.

After a couple of minutes of my laptop warming up, I got to a fresh copy of Helix in a debug folder. I then set it up with a phrase with some quotation marks in it, and then selected the whole line with `x`, and then got ready to replace it all with an extra newline at the end:

![0  before](https://github.com/user-attachments/assets/e0ef1e5d-341c-4ab7-a0ea-a8f3546bd8d2)

And…I got that extra newline dumped into my document, as it should:

![1  extra newline](https://github.com/user-attachments/assets/dfb9ccb9-20bf-48fe-8060-4233626a00fc)

----

I tried running `quote-educator` (without the `-n`), and it does the right thing (not pictured).

----

I also tried `W` and `E` for the phrase `I'd` and both do the right thing.

Guess I just need to sit tight and wait for what's in `master` to get released. Thanks!
